### ISSUE TITLE
General Single Crystal Inelastic Material - cont'd

### ISSUE THREAD

**Issue by willend** [mccode-dev/McCode#1251]
The right place to continue McStas-oriented discussion starting from https://github.com/mctools/ncrystal/discussions/87, that started off like this post by @thallor1:

> Hello - I was wondering if a feature exists within Ncrystal to simulate scattering for a general four dimensional (Qx, Qy, Qz, Energy) space, assuming that the cross section has already been calculated for some grid of points within this space. The ultimate goal is to use this in McStas, but the current implementation (See [here](http://www.mcstas.org/download/components/2.7.1/contrib/Single_crystal_inelastic.html)) is non functional. However, the described functionality is exactly what I'm looking for and I was wondering if this is possible using Ncrystal.

**Comment by thallor1** [comment]
Continuing discussion from the Ncrystal thread - 

I have tried spot-sample but it doesn't exactly suit my use case. It does not allow for a crystal mosaicity and does not support full single crystal scattering, only a defined two-theta so it behaves more like a powder. 

The current workaround that I am using is to simply lower the outgoing neutron kf length in the standard single_crystal component, which seems to work. The problem with this is that I need to account for the change in the scattering angle, which I do by adjusting the HKL indices in the .hkl file. I think that I'm essentially doing the same thing that single_crystal_inelastic is.

**Comment by farhi** [comment]
Hi all,

just my 2c. There exists an inelastic SX component, initially written by Duc Le, and cleaned/optimized/pushed into McStas:
- https://github.com/McStasMcXtrace/McCode/blob/master/mcstas-comps/contrib/Single_crystal_inelastic.comp

It is already part of McStas (2.x), but not ported for GPU yet (mccode-3).

**Comment by willend** [comment]
@farhi yup, actually @thallor1 specifically mentions that comp in the original post on the mcpl list (https://github.com/mctools/ncrystal/discussions/87)

It looks like it could need some further brushing up

**Comment by willend** [comment]
@mads-bertelsen any word from your US connection on potential collaboration in this direction?

**Comment by mads-bertelsen** [comment]
It sadly seems the funding was suddenly cut for that position and the work was not finished.

There is something that should be functional here: https://github.com/thallor1/pyMACS/blob/main/pyMACS/UNION%20MACS%20Models/UNION%20MACS%20Base/Sqw4_process.comp

**Comment by willend** [comment]
Very cool that you found a comp! I will clone the repo to mccode-dev and later set up a dedicated branch here + PR to include some bits.

**Comment by thallor1** [comment]
Hello - would still be happy to help occasionally with this, there are
still a number of people at NIST that would be interested in seeing this be
more functional. Not sure how much I can help, but it's more or less a port
of the single_crystal_inelastic.comp to union. All of the work that I did
in that direction should be in that github repo, with some examples in the
documentation <https://thallor1.github.io/pyMACS/>.
